File: targets/widget/Fetcher.swift
import Foundation

struct NoBody: Encodable {}

enum HTTPMethod: String {
  case GET = "GET"
  case POST = "POST"
  case PUT = "PUT"
  case PATCH = "PATCH"
  case DELETE = "DELETE"
}

struct FetchParams<T: Encodable> {
  let method: HTTPMethod
  let url: String
  let baseUrl: String?
  let connection: Connection
  let body: T?
  
  init(method: HTTPMethod, url: String, connection: Connection, body: T, baseUrl: String? = nil) {
    self.method = method
    self.url = url
    self.connection = connection
    self.body = body
    self.baseUrl = baseUrl
  }
  
  init(method: HTTPMethod, url: String, connection: Connection, baseUrl: String? = nil) {
    self.method = method
    self.url = url
    self.connection = connection
    self.body = nil
    self.baseUrl = baseUrl
  }
}

private func fetch<T: Encodable>(params: FetchParams<T>, completion: @escaping (Result<Data, Error>) -> Void) {
  if (!params.url.starts(with: "/")) {
    return completion(.failure(NSError(domain: "InvalidUrl", code: 0, userInfo: [NSLocalizedDescriptionKey: "URL should start with /"])))
  }
  
  let fullUrlString = params.baseUrl != nil ? "\(params.baseUrl ?? "")\(params.url)" : "https://api.vercel.com\(params.url)"
  
  guard let fullUrl = URL(string: fullUrlString) else {
    return completion(.failure(NSError(domain: "InvalidURL", code: 0, userInfo: [NSLocalizedDescriptionKey: "Invalid URL"])))
  }
  
  var request = URLRequest(url: fullUrl)
  
  request.httpMethod = params.method.rawValue
  request.addValue("application/json", forHTTPHeaderField: "Accept")
  request.addValue("Bearer \(params.connection.apiToken)", forHTTPHeaderField: "Authorization")
  
  if let data = params.body {
    let jsondata = try? JSONEncoder().encode(data)
    request.httpBody = jsondata
  }
  
  let task = widgetSession.dataTask(with: request) { data, response, error in
    if let error = error {
      completion(.failure(error))
      return
    }
    
    guard let httpResponse = response as? HTTPURLResponse else {
      return completion(.failure(NSError(domain: "InvalidResponse", code: 0, userInfo: [NSLocalizedDescriptionKey: "Invalid response"])))
    }
    
    if !(200...299).contains(httpResponse.statusCode) {
      let error = NSError(domain: "HTTPError", code: httpResponse.statusCode, userInfo: [NSLocalizedDescriptionKey: "HTTP Error: \(httpResponse.statusCode)"])
      
      if let data = data, let errorString = String(data: data, encoding: .utf8) {
        print("Error Response Body: \(errorString)")
      }
      
      return completion(.failure(error))
    }
    
    guard let data = data else {
      return completion(.failure(NSError(domain: "NoData", code: 0, userInfo: [NSLocalizedDescriptionKey: "No data received"])))
    }
    
    completion(.success(data))
  }
  
  task.resume()
}

func httpRequest<T: Decodable, K: Encodable>(params: FetchParams<K>) async throws -> T {
  try await withCheckedThrowingContinuation { continuation in
    fetch(params: params) { result in
      switch result {
      case .success(let data):
        do {
          let decoder = JSONDecoder()
          let decodedResult = try decoder.decode(T.self, from: data)
          
          continuation.resume(returning: decodedResult)
        } catch {
          continuation.resume(throwing: error)
        }
      case .failure(let error):
        continuation.resume(throwing: error)
      }
    }
  }
}

/// One session for the whole extension. The default 60 s request timeout can eat the entire
/// WidgetKit reload budget, and a `URLSession` created per request is never invalidated.
let widgetSession: URLSession = {
  let configuration = URLSessionConfiguration.default
  configuration.timeoutIntervalForRequest = 15
  return URLSession(configuration: configuration)
}()

// MARK: - Favicons
//
// Vercel removed `vercel.com/api/v0/deployments/<id>/favicon`, and raw deployment URLs sit behind
// Deployment Protection (every path answers with a 200 HTML login page). The icon is therefore
// resolved from the production domain: the page's declared icon first, then the conventional
// paths, accepting only image responses. Mirrors `lib/favicon.ts` in the app.

private let faviconCacheMaxAge: TimeInterval = 24 * 60 * 60
private let faviconFallbackPaths = ["/favicon.ico", "/favicon.png", "/apple-touch-icon.png"]
// Deployment Protection redirects to vercel.com/sso-api and then to vercel.com/login
private let protectionHost = "vercel.com"
private let protectionMarkers = ["/sso-api", "_vercel_sso"]

private func faviconFileURL(projectId: String) -> URL? {
  FileManager.default.containerURL(forSecurityApplicationGroupIdentifier: appGroupName)?
    .appendingPathComponent("favicon-\(projectId)")
}

/// Path of the cached icon when it is younger than a day.
private func cachedFaviconPath(projectId: String) -> String? {
  guard let url = faviconFileURL(projectId: projectId),
        let attributes = try? FileManager.default.attributesOfItem(atPath: url.path),
        let modified = attributes[.modificationDate] as? Date,
        Date().timeIntervalSince(modified) < faviconCacheMaxAge else {
    return nil
  }
  
  return url.path
}

// `UIImage` cannot render SVG, and a protection page is text/html
private func isRenderableImage(_ response: URLResponse) -> Bool {
  guard let http = response as? HTTPURLResponse,
        (200...299).contains(http.statusCode),
        let type = http.value(forHTTPHeaderField: "Content-Type")?.lowercased() else {
    return false
  }
  
  return type.hasPrefix("image/") && !type.contains("svg")
}

private func isProtectionPage(response: HTTPURLResponse, html: String) -> Bool {
  let finalHost = response.url?.host?.lowercased() ?? ""
  
  return finalHost == protectionHost
    || finalHost.hasSuffix(".\(protectionHost)")
    || protectionMarkers.contains(where: { html.contains($0) })
}

private func attribute(_ name: String, in tag: String) -> String? {
  guard let regex = try? NSRegularExpression(pattern: "\\b\(name)=[\"']([^\"']+)[\"']", options: .caseInsensitive),
        let match = regex.firstMatch(in: tag, range: NSRange(tag.startIndex..., in: tag)),
        let range = Range(match.range(at: 1), in: tag) else {
    return nil
  }
  
  return String(tag[range])
}

/// Icon hrefs declared in the page head, SVGs skipped. Touch icons last: they are often 1 MB PNGs.
func extractIconHrefs(from html: String) -> [String] {
  guard let tagRegex = try? NSRegularExpression(pattern: "<link\\b[^>]*>", options: .caseInsensitive) else {
    return []
  }
  
  var touchIcons: [String] = []
  var icons: [String] = []
  
  for match in tagRegex.matches(in: html, range: NSRange(html.startIndex..., in: html)) {
    guard let tagRange = Range(match.range, in: html) else { continue }
    let tag = String(html[tagRange])
    
    guard let rel = attribute("rel", in: tag)?.lowercased(),
          rel.contains("icon"), !rel.contains("mask-icon"),
          let href = attribute("href", in: tag) else {
      continue
    }
    
    if href.lowercased().split(separator: "?").first?.hasSuffix(".svg") == true { continue }
    
    if rel.contains("apple-touch-icon") {
      touchIcons.append(href)
    } else {
      icons.append(href)
    }
  }
  
  return icons + touchIcons
}

private func resolveHref(_ href: String, base: String) -> URL? {
  if href.hasPrefix("http://") || href.hasPrefix("https://") { return URL(string: href) }
  if href.hasPrefix("//") { return URL(string: "https:\(href)") }
  return URL(string: href.hasPrefix("/") ? "\(base)\(href)" : "\(base)/\(href)")
}

/// Downloads the favicon of `host` into the app group container and returns its path.
private func downloadWebsiteFavicon(host: String, projectId: String) async -> String? {
  let base = "https://\(host)"
  var candidates: [URL] = []
  
  if let homepage = URL(string: base),
     let (data, response) = try? await widgetSession.data(from: homepage),
     let http = response as? HTTPURLResponse,
     (200...299).contains(http.statusCode),
     http.value(forHTTPHeaderField: "Content-Type")?.lowercased().contains("text/html") == true,
     let html = String(data: data, encoding: .utf8) {
    if isProtectionPage(response: http, html: html) { return nil }
    
    candidates = extractIconHrefs(from: html).prefix(3).compactMap { resolveHref($0, base: base) }
  }
  
  candidates += faviconFallbackPaths.compactMap { URL(string: "\(base)\($0)") }
  
  guard let fileURL = faviconFileURL(projectId: projectId) else { return nil }
  
  for candidate in candidates {
    guard let (data, response) = try? await widgetSession.data(from: candidate),
          isRenderableImage(response), !data.isEmpty else {
      continue
    }
    
    // atomic: several widgets may refresh the same project at once
    if (try? data.write(to: fileURL, options: .atomic)) != nil {
      return fileURL.path
    }
  }
  
  return nil
}

/// Favicon path for a project, cached for a day. Pass the production domain when the caller
/// already has it (from `production-deployment`), otherwise it is fetched.
func fetchProjectFavicon(project: ProjectListItem, productionDomain: String? = nil) async -> String? {
  if let cached = cachedFaviconPath(projectId: project.id) { return cached }
  
  var host = productionDomain
  if host == nil {
    host = try? await fetchProductionDeployment(
      connection: project.connection,
      connectionTeam: project.connectionTeam,
      projectId: project.id
    ).domain?.name
  }
  
  guard let host, !host.isEmpty else { return nil }
  
  return await downloadWebsiteFavicon(host: host, projectId: project.id)
}

File: targets/widget-android/src/main/java/package_name/Fetcher.kt
package com.revcel.mobile

import ProjectListItem
import android.content.Context
import java.net.HttpURLConnection
import java.net.URL
import kotlinx.coroutines.Dispatchers
import kotlinx.coroutines.withContext
import java.io.ByteArrayOutputStream
import com.google.gson.Gson
import expo.modules.widgetkit.Connection
import java.io.File
import java.io.OutputStreamWriter

enum class HTTPMethod(val value: String) {
    GET("GET"),
    POST("POST"),
    PUT("PUT"),
    PATCH("PATCH"),
    DELETE("DELETE")
}

data class FetchParams(
    val method: HTTPMethod,
    val url: String,
    val connection: Connection,
    val body: String? = null,
    val baseUrl: String? = null
)

// MARK: favicons
//
// Vercel removed `vercel.com/api/v0/deployments/<id>/favicon`, and raw deployment URLs sit behind
// Deployment Protection (every path answers with a 200 HTML login page). The icon is therefore
// resolved from the production domain: the page's declared icon first, then the conventional
// paths, accepting only image responses. Mirrors `lib/favicon.ts` in the app.

private const val FAVICON_CACHE_MAX_AGE_MS = 24L * 60 * 60 * 1000
private val FAVICON_FALLBACK_PATHS = listOf("/favicon.ico", "/favicon.png", "/apple-touch-icon.png")
// Deployment Protection redirects to vercel.com/sso-api and then to vercel.com/login
private const val PROTECTION_HOST = "vercel.com"
private val PROTECTION_MARKERS = listOf("/sso-api", "_vercel_sso")
private val LINK_TAG_REGEX = Regex("<link\\b[^>]*>", RegexOption.IGNORE_CASE)
private val REL_ATTR_REGEX = Regex("\\brel=[\"']([^\"']+)[\"']", RegexOption.IGNORE_CASE)
private val HREF_ATTR_REGEX = Regex("\\bhref=[\"']([^\"']+)[\"']", RegexOption.IGNORE_CASE)

private fun openWebsiteConnection(url: String): HttpURLConnection =
    (URL(url).openConnection() as HttpURLConnection).apply {
        requestMethod = HTTPMethod.GET.value
        instanceFollowRedirects = true
        connectTimeout = 15000
        readTimeout = 15000
        // arbitrary websites, some of which reject an empty user agent
        setRequestProperty("User-Agent", "Mozilla/5.0 (compatible; RevcelWidget)")
    }

private fun faviconFile(context: Context, projectId: String) = File(context.filesDir, "favicon-$projectId")

/** Path of the cached icon when it is younger than a day. */
private fun cachedFaviconPath(context: Context, projectId: String): String? {
    val file = faviconFile(context, projectId)
    val isFresh = file.exists() && file.length() > 0 &&
        System.currentTimeMillis() - file.lastModified() < FAVICON_CACHE_MAX_AGE_MS
    return if (isFresh) file.path else null
}

// BitmapFactory cannot render SVG, and a protection page is text/html
private fun isRenderableImage(connection: HttpURLConnection): Boolean {
    val type = connection.contentType?.lowercase() ?: return false
    return connection.responseCode in 200..299 && type.startsWith("image/") && !type.contains("svg")
}

/** Downloads `url` into `target` when it is an image, atomically via a temp file. */
private fun downloadImage(url: String, target: File): Boolean {
    val connection = try { openWebsiteConnection(url) } catch (e: Exception) { return false }
    var temp: File? = null
    try {
        connection.connect()
        if (!isRenderableImage(connection)) return false

        // several widgets may refresh the same project at once, never write the target directly
        val tempFile = File.createTempFile(target.name, null, target.parentFile)
        temp = tempFile
        connection.inputStream.use { input -> tempFile.outputStream().use { output -> input.copyTo(output) } }

        return tempFile.length() > 0 && tempFile.renameTo(target)
    } catch (e: Exception) {
        return false
    } finally {
        temp?.takeIf { it.exists() }?.delete()
        connection.disconnect()
    }
}

/** The page html and the host that finally served it (after redirects). */
private fun fetchHtml(url: String): Pair<String, String>? {
    val connection = try { openWebsiteConnection(url) } catch (e: Exception) { return null }
    try {
        connection.connect()
        val type = connection.contentType?.lowercase() ?: ""
        if (connection.responseCode !in 200..299 || !type.contains("text/html")) return null
        val html = connection.inputStream.bufferedReader().use { it.readText() }
        return html to (connection.url.host?.lowercase() ?: "")
    } catch (e: Exception) {
        return null
    } finally {
        connection.disconnect()
    }
}

private fun isProtectionPage(finalHost: String, html: String): Boolean =
    finalHost == PROTECTION_HOST ||
        finalHost.endsWith(".$PROTECTION_HOST") ||
        PROTECTION_MARKERS.any { html.contains(it) }

/** Icon hrefs declared in the page head, SVGs skipped. Touch icons last: they are often 1 MB PNGs. */
fun extractIconHrefs(html: String): List<String> {
    val touchIcons = mutableListOf<String>()
    val icons = mutableListOf<String>()

    for (tag in LINK_TAG_REGEX.findAll(html)) {
        val rel = REL_ATTR_REGEX.find(tag.value)?.groupValues?.get(1)?.lowercase() ?: continue
        val href = HREF_ATTR_REGEX.find(tag.value)?.groupValues?.get(1) ?: continue

        if (!rel.contains("icon") || rel.contains("mask-icon")) continue
        if (href.lowercase().substringBefore('?').endsWith(".svg")) continue

        if (rel.contains("apple-touch-icon")) touchIcons.add(href) else icons.add(href)
    }

    return icons + touchIcons
}

private fun resolveHref(base: String, href: String): String = when {
    href.startsWith("http://") || href.startsWith("https://") -> href
    href.startsWith("//") -> "https:$href"
    href.startsWith("/") -> "$base$href"
    else -> "$base/$href"
}

/** Downloads the favicon of `host` and returns the file path. Refuses Vercel's protection page. */
private suspend fun downloadWebsiteFavicon(context: Context, host: String, projectId: String): String? =
    withContext(Dispatchers.IO) {
        val base = "https://$host"
        val candidates = mutableListOf<String>()

        val page = fetchHtml(base)
        if (page != null) {
            val (html, finalHost) = page
            if (isProtectionPage(finalHost, html)) return@withContext null
            candidates += extractIconHrefs(html).take(3).map { resolveHref(base, it) }
        }
        candidates += FAVICON_FALLBACK_PATHS.map { "$base$it" }

        val target = faviconFile(context, projectId)
        for (candidate in candidates) {
            if (downloadImage(candidate, target)) return@withContext target.path
        }

        null
    }

/**
 * Favicon path for a project, cached for a day. Pass the production domain when the caller
 * already has it (from `production-deployment`), otherwise it is fetched.
 */
suspend fun fetchProjectFavicon(context: Context, project: ProjectListItem, productionDomain: String? = null): String? {
    cachedFaviconPath(context, project.id)?.let { return it }

    val host = productionDomain ?: try {
        fetchProductionDeployment(project.connection, project.connectionTeam, project.id).domain?.name
    } catch (e: Exception) {
        null
    }
    if (host.isNullOrEmpty()) return null

    return downloadWebsiteFavicon(context, host, project.id)
}

suspend fun fetch(params: FetchParams): ByteArray = withContext(Dispatchers.IO) {
    val isPOSTRequest = params.body !== null && params.method === HTTPMethod.POST

    if (!params.url.startsWith("/")) {
        throw Exception("URL should start with /")
    }

    val fullUrlString = if (params.baseUrl !== null) "${params.baseUrl}${params.url}" else "https://api.vercel.com${params.url}"
    val url = URL(fullUrlString)

    val connection = (url.openConnection() as HttpURLConnection).apply {
        requestMethod = params.method.value
        setRequestProperty("User-Agent", "")
        setRequestProperty("Accept", "application/json")
        setRequestProperty("Authorization", "Bearer ${params.connection.apiToken}")
        connectTimeout = 15000
        readTimeout = 15000
    }

    try {
        if (isPOSTRequest) {
            connection.doOutput = true
            OutputStreamWriter(connection.outputStream).use { writer ->
                writer.write(params.body)
                writer.flush()
            }
        }

        connection.connect()
        val responseCode = connection.responseCode

        if (responseCode !in 200..299) {
            val errorMsg = connection.errorStream?.bufferedReader()?.use { it.readText() }
                ?: "HTTP Error: $responseCode"

            throw Exception("HTTP Error: $responseCode. $errorMsg")
        }

        connection.inputStream.use { input ->
            val buffer = ByteArrayOutputStream()
            val data = ByteArray(1024)
            var nRead: Int
            while (input.read(data, 0, data.size).also { nRead = it } != -1) {
                buffer.write(data, 0, nRead)
            }
            buffer.toByteArray()
        }
    } finally {
        connection.disconnect()
    }
}

suspend inline fun <reified T> httpRequest(params: FetchParams): T {
    val data = fetch(params)
    val json = String(data, Charsets.UTF_8)

    return Gson().fromJson(json, T::class.java)
}
